### DISCUSSION TITLE
How can I check for close notify from the server before sending the request

### DISCUSSION THREAD

**Discussion by SergeyKiwi** [openssl/openssl#21861] [Q&A] [answered]
I'm using OpenSSL 3.0.9. I'm trying to write C++ code that does this:

```
openssl s_client -connect <host>:<port> -tls1_2 -no_ticket -CAfile <file> 
```

When the server timeout ends, it sends a "close notify" notification and the application closes responding.

However, in my code, the only way to get the message is `BIO_read()`, which initializes the SSL connection state. And then use `if (SSL_get_shutdown(ssl) == SSL_RECEIVED_SHUTDOWN)`. If bio is blocking, then `BIO_read` will cause the application to hang on the server timeout. And I can't `BIO_read()` before `BIO_write()`.

How can I find out about the connection status before sending the request?

How can I initialize the connection status? 

Is it possible to do this with blocking bio?

[here is the full question](https://stackoverflow.com/q/76835708/22338524)

**Comment by mattcaswell** [comment, marked as answer]
Your suggested answer in that stack overflow question is the correct approach for using a blocking socket (or alternatively to `select` you can use `poll`/`epoll`/whatever). There is a risk that the peer only sends "half" a record in which case `select` will indicate that the socket is readable and then the application will read the first half of the record, but then block waiting for the rest of the record to arrive - but that doesn't normally happen with a well behaved peer.

Alternatively, you can of course use a non-blocking socket.

**Reply by SergeyKiwi** [reply]
Thank you for your reply. In my case, there is no risk.

**Comment by SergeyKiwi** [comment]
My suggested answer is incorrect when the close notify received after checking before sending the request. Then, my client sends the request. I see it in Wireshark. However, the server doesn't receive it. Therefore, my read method returns 0. I'm already getting these errors in practice. It doesn't feel right to me. I'm trying to understand the openssl s_client source code, but it's hard.

**Reply by mattcaswell** [reply]
This race condition will always exist. There will always be a time lag between the server sending something and it arriving at the other end. So even if you send something to the server and have not yet received a close_notify there is no way to know whether you will immediately receive the close_notify that was already in-flight.

**Reply by SergeyKiwi** [reply]
I agree with you. I will follow up on closure messages after submitting the request and submit the request again. Thank you for your attention
